### ISSUE TITLE
Clock keeps spinning after all tasks completed

### ISSUE THREAD

**Issue by bwims** [joonaspaakko/audible-library-extractor#130]
Hi,

My account is on audible.co.uk , if that makes a difference?

All tasks completed but the process then hangs with the clock spinning and gallery never appears.

I have disabled ad blocker and cookie trackers but no joy.

B.

**Comment by bwims** [comment]
After reading further, I thought this might be helpful

![image](https://github.com/joonaspaakko/audible-library-extractor/assets/18516073/20de2215-744a-45d6-b7b0-d3b1b7e381fd)

**Comment by joonaspaakko** [comment]
The good news is that I think I know where the issue is coming. The bad news is you have to wait for me to make a new build that fixes it. I'll try to fix it later today.

**Comment by bwims** [comment]
Great!  Thanks!

**Comment by joonaspaakko** [comment]
Could you try this dev build?

**Installation instructions:**
1. [Download the 0.2.10 draft](https://github.com/joonaspaakko/audible-library-extractor/raw/master/audible-library-extractor-0.2.10.zip) 
2. Go to the address: `chrome:extensions` (or from the menu in top right: `⋮ > Extensions > Manage extensions`)
3. Enable the developer mode from the top right
4. Drag the downloaded .zip file into the browser window to install

**Comment by bwims** [comment]
Hi,

Sorry for the delay in replying, and many thanks for allowing me to try out
your beta!

It is really impressive, and does the main thing that I wanted which is
export a spreadsheet of my library.

There are a couple of little niggles, that might only apply because I am
using an old version of Excel.

1)  in text descriptions, the character code results in the apostrophe
being represented as this string:  "â€™"  I think this is just a matter of
encoding and it's my problem really.

2) what looks like it could be a real problem is that sometimes an infinity
symbol is being inserted as a book number and seems to only occur when it
is a numberless part of a series,
for example "Memory's Legion" from the "Expanse" series, and "Tales of
Dune: Expanded Edition" , [art of the "Dune Saga".

I believe the infinity symbol is what "âˆž" encodes to in UTF-8.
Interestingly, your gallery displays the book number of the series in
brackets, except when it is an infinity symbol, so I assume it is some kind
of flag.

Not very important niggles!

Many thanks, and keep me in mind to try out anything further for you.

Best regards

Brian

**Comment by joonaspaakko** [comment]
@bwims, I think the issue might be importing the CSV with Excel's default encoding Western 1252. At least that's how I was able to get `â€™` and `âˆž` to show up in my text editor. If I open the CSV with UTF-8 encoding it looks normal.

![image](https://github.com/joonaspaakko/audible-library-extractor/assets/1164476/d9c5d16d-eaf9-4224-8076-73ef30d4d80c)

Or if that doesn't work, I think [sheets.google.com](https://sheets.google.com/) would potentially work better. I believe it prefers UTF-8. Funny enough, looking into this I found many articles for Excel users saying something like "If you see weird characters in your CSV file in Excel, import the file to Google Sheets because it opens files with UTF-8 encoding, download it as an Excel file and now it shows up properly in Excel", which seems a bit silly for many reasons; you could just use Google Sheets and Excel also has its own import encoding option `File Origin` (as shown in the picture).

- - - - 

**What concerns me a little is:**

> Interestingly, your gallery displays the book number of the series in
brackets, except when it is an infinity symbol, so I assume it is some kind
of flag.

By brackets do you mean something like `[1]` or `[1,2,3]` or maybe `["1", "2"]`? This can happen sometimes when rendering an array as such and definitely isn't on purpose. It could've happened with the changes that I made on the v.0.2.10 draft, but I'm not seeing any brackets around book numbers with my test data. I would be interested to see a screenshot of what you're seeing.

**Comment by bwims** [comment]
All points on encoding well taken.

Regarding book number, brackets aren't the problem.  
Ex 1

This book has the value "âˆž" (∞) in the field "Book Numbers" of the CSV export 

Note that there is no book number in brackets


![image](https://github.com/joonaspaakko/audible-library-extractor/assets/18516073/55915052-81bc-4832-b89e-04826671dbb1)

However, this book has the number 8 in the field "Book Numbers" of the CSV export 
and displays (book 8) in brackets.

![image](https://github.com/joonaspaakko/audible-library-extractor/assets/18516073/916baf27-2251-4947-bec8-54a012081371)


But it isn't a big deal - just a curiosity!

Thanks!

B.

**Comment by bwims** [comment]
PS your text import example worked wonderfully!  ( and it is the infinity symbol!)

**Comment by bwims** [comment]
PPS,  Would it be possible to include date of purchase?  Or is that another dataset?

**Comment by joonaspaakko** [comment]
Oh I see... That's a relief. Apparently I think of `()` as parenthesis and `[]` as brackets (or square brackets), hence the confusion.

So Yea, the brackets are there on purpose. They are supposed to further indicate that this is the book number and not part of the series name all the while not trying to break the listing of series names, since some books are part of multiple series.

Audible does this:
[This is a series](#), book 1, [This is another series the book is part of](#), book 6

**I just don't like it so the Galler shows it like:**

[This is a series](#) (book 1), [This is another series the book is part of](#) (book 6)

My gallery, my rules... I guess :)

**Comment by joonaspaakko** [comment]
> PPS,  Would it be possible to include date of purchase?  Or is that another dataset?

Tl;Dr yes, it's on my to-do list.

The website library as well as the store page used to display date added a few years ago, but they took it away from both. when they modernized the library page. 

For v.0.2.9 I was supposed to add an extraction for the purchase history pages, but too many issues piled up and I didn't have the time, given that the time was ticking and the extraction was broken for everyone. Now it's back on the to-do list but I won't make promises on when or even if it will actually be added, only that I will see what I can do... At some point.

**Comment by bwims** [comment]
Re Book number,  I had assumed what you say.  I guess I said it in a confusing way.  My whole point was to conjecture that the infinity symbol is there to flag a book in a series but without a book number.

I really appreciate your efforts.  I feel very angry with Amazon for their arbitrary changes and not providing the data extraction facilities in an easy way.   They used to have an indicator to show when a book had been downloaded or not, but no longer.

It must be a real labour of love for you!

Thanks!

B.

**Comment by joonaspaakko** [comment]
I hadn't even noticed the downloaded status was taken away... Good to know.

It is a labor of lover/hate/necessity. Audible has caught up quite a bit in recent years, but I still use the gallery on my phone for browsing my library to find the next book and the Audible mobile app just for listening. 

Though I recognize the wallpaper creator (in the gallery menu "extension tools") is very niche and not at all necessary, I'm very fond of it too. I've made animated wallpapers and screensavers for all my computers. Admittedly, I've never really used the CSV export myself and it was pretty much just an afterthought.

**Comment by bwims** [comment]
It was a great afterthought!  

Thanks for all you help!

**Comment by TheSidXXX** [comment]
I was having the same issue, and the 2.10 version linked above solved it for me.  Weird glitch, Audible and the tool both say I have 1738 books, but once the gallery was generated it only shows 1734.  Not sure of an easy way to figure out which 4 aren't in there, but if I do figure it out, I'll let you know.

**Comment by joonaspaakko** [comment]
@TheSidXXX, off the top of my head  my best guess is that you have 4 podcasts on the "podcasts" page.

> It goes through the gallery on the "All titles" tab just because there wasn't a reason not to. Now that it includes podcasts it really should show the counters somehow separately to avoid confusion.

If you don't think that is the reason, maybe try extracting again (without removing any data beforehand).

**Comment by TheSidXXX** [comment]
> @TheSidXXX, off the top of my head my best guess is that you have 4 podcasts on the "podcasts" page.
> 
> > The extraction goes through the gallery on the "All titles" tab just because there wasn't a reason not to. Now that it includes podcasts it really should show the counters somehow separately to avoid confusion.
> 
> If you don't think that is the reason, maybe try extracting again (without removing any data beforehand).

That looks like it!  I have 4 podcasts on there.

Another weird issue, I thought everything worked, but when I go to the series page, nothing loads.  I can see some thumbnails in the background, but the actual list of series doesn't load.  Authors, narrators, publisher, category all work, but series doesn't.  Since there are series screenshots in this thread, I'm guessing it's something on my end.

**Comment by joonaspaakko** [comment]
@TheSidXXX, could you export your raw data from the extraction settings and send that to me? You can email it to me, though it doesn't contain anything more than the website data, it's just formatted differently. My email is in my Github profile.

**Comment by TheSidXXX** [comment]
> @TheSidXXX, could you export your raw data from the extraction settings and send that to me? You can email it to me, though it doesn't contain anything more than the website data, it's just formatted differently. My email is in my Github profile.

I removed all extracted data and reran it, and series worked this time.  I can still send if you want, but it looks like it's been resolved.

**Comment by joonaspaakko** [comment]
@TheSidXXX, no need in that case, but if it happens again... 

I'm maybe planning to make a quick new release soon because of this issue, but I'd rather not just swap it with another issue, so if you or I guess anyone else gets the series page is blank issue again, I'd appreciate the raw data.

**Comment by TheSidXXX** [comment]
@joonaspaakko I still have the zip file of the gallery that had the issue, would that be helpful?
[ALE-gallery (7).zip](https://github.com/joonaspaakko/audible-library-extractor/files/12771807/ALE-gallery.7.zip)

**Comment by joonaspaakko** [comment]
That should do it, thanks!
